lib/linked_list.rb

```diff
@@ -19,48 +19,105 @@ def initialize
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
     def add_first(value)
-      raise NotImplementedError
+        current_node = Node.new(value, @head)
+        @head = current_node
     end
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
     def search(value)
-      raise NotImplementedError
-    end
+      temp = @head
+      until temp.nil?
+        if temp.data == value
+          return true
+        end
+        temp = temp.next
+    end
+    return false 
+  end
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max
-      raise NotImplementedError
+      max = nil
+      current = @head 
+      until current.nil?
+        if max.nil? || current.data > max
+          max = current.data
+        end
+        current = current.next
+      end
+      return max
     end
 
     # method to return the min value in the linked list
     # returns the data value and not the node
     def find_min
-      raise NotImplementedError
+      min = nil
+      current = @head 
+      until current.nil?
+        if min.nil? || current.data < min
+          min = current.data
+        end
+        current = current.next
+      end
+      return min
     end
 
 
     # method that returns the length of the singly linked list
     def length
-      raise NotImplementedError
+      count = 0
+      current = @head 
+      until current.nil?
+        count += 1 
+        current = current.next
+      end
+      return count
     end
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns nil if there are fewer nodes in the linked list than the index value
     def get_at_index(index)
-      raise NotImplementedError
+      count = 0
+      current = @head
+
+      until current.nil?
+        count += 1
+        if index == count
+          return current.data
+        end
+
+        current = current.next
+      end
+      return nil
     end
+  # end
 
     # method to print all the values in the linked list
     def visit
-      raise NotImplementedError
+      print_all = nil 
+      current = @head
+      until current.nil?
+        print current.data
+        current = current.next
+      end
     end
 
     # method to delete the first node found with specified value
     def delete(value)
-      raise NotImplementedError
+      if @head.nil?
+        return nil 
+      end
+
+      previous = @head 
+      current = @head 

```

Comment:
You should also put in a check to make sure that the node we're deleting is not the head.